pkg/build/builder/cmd/builder.go

```diff
@@ -21,162 +21,175 @@ import (
+	}
+	cfg.buildsClient = osClient.Builds(cfg.build.Namespace)
+
+	return cfg, nil
+}
+
+func (c *builderConfig) setupGitEnvironment() ([]string, error) {
+
+	gitSource := c.build.Spec.Source.Git
+
+	// For now, we only handle git. If not specified, we're done
+	if gitSource == nil {
+		return []string{}, nil
+	}
+
+	sourceSecret := c.build.Spec.Source.SourceSecret
+	gitEnv := []string{"GIT_ASKPASS=true"}
+	// If a source secret is present, set it up and add its environment variables
+	if sourceSecret != nil {
+		// TODO: this should be refactored to let each source type manage which secrets
+		//   it accepts
+		sourceURL, err := git.ParseRepository(gitSource.URI)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse build URL: %s", gitSource.URI)
+		}
+		scmAuths := scmauth.GitAuths(sourceURL)
+
+		// TODO: remove when not necessary to fix up the secret dir permission
+		sourceSecretDir, err := fixSecretPermissions(c.sourceSecretDir)
+		if err != nil {
+			return nil, fmt.Errorf("cannot fix source secret permissions: %v", err)
+		}
+
+		secretsEnv, overrideURL, err := scmAuths.Setup(sourceSecretDir)
+		if err != nil {
+			return nil, fmt.Errorf("cannot setup source secret: %v", err)
+		}
+		if overrideURL != nil {
+			c.build.Annotations[bld.OriginalSourceURLAnnotationKey] = gitSource.URI
+			gitSource.URI = overrideURL.String()
+		}
+		gitEnv = append(gitEnv, secretsEnv...)
+	}
+	if len(gitSource.HTTPProxy) > 0 {
+		gitEnv = append(gitEnv, fmt.Sprintf("HTTP_PROXY=%s", gitSource.HTTPProxy))
+		gitEnv = append(gitEnv, fmt.Sprintf("http_proxy=%s", gitSource.HTTPProxy))
+	}
+	if len(gitSource.HTTPSProxy) > 0 {
+		gitEnv = append(gitEnv, fmt.Sprintf("HTTPS_PROXY=%s", gitSource.HTTPSProxy))
+		gitEnv = append(gitEnv, fmt.Sprintf("https_proxy=%s", gitSource.HTTPSProxy))
 	}
-	osClient, err := client.New(config)
+	return bld.MergeEnv(os.Environ(), gitEnv), nil
+}
+
+// execute is responsible for running a build
+func (c *builderConfig) execute(b builder) error {
+
+	gitEnv, err := c.setupGitEnvironment()
 	if err != nil {
-		glog.Fatalf("Error obtaining OpenShift client: %v", err)
+		return err
 	}
-	buildsClient := osClient.Builds(build.Namespace)
+	gitRepo := git.NewRepositoryWithEnv(gitEnv)
 
-	if err = b.Build(dockerClient, endpoint, buildsClient, &build); err != nil {
-		glog.Fatalf("Build error: %v", err)
+	if err := b.Build(c.dockerClient, c.dockerEndpoint, c.buildsClient, c.build, gitRepo); err != nil {
+		return fmt.Errorf("build error: %v", err)
 	}
 
-	if build.Spec.Output.To == nil || len(build.Spec.Output.To.Name) == 0 {
+	if c.build.Spec.Output.To == nil || len(c.build.Spec.Output.To.Name) == 0 {
 		glog.Warning("Build does not have an Output defined, no output image was pushed to a registry.")
 	}
 
+	return nil
 }
 
 // fixSecretPermissions loweres access permissions to very low acceptable level
 // TODO: this method should be removed as soon as secrets permissions are fixed upstream
```

Comment:
This is not fixed. Requires a solution to this issue: https://github.com/kubernetes/kubernetes/issues/4789